Various cleanups in the Inliner

lrytz opened the pull request.

lrytz committed:
Rename filter to withFilter in RightBiasedEither

lrytz committed:
Don't create inline requests for callsites that cannot be inlined

When traversing the call graph and collecting inline reqeusts, rule
out callsites that we already know cannot be inlined.

Note that we cannot perform all necessary checks already at this
stage: checks that depend on the callee body (the inlined code) are
deferred until the callsite is actually inlined. The reason is that
the code may change. Example:

    @inline final def f = try 1 catch { case _: Throwable => 2 }
    @inline final def g = f
    def t = println(g)

When collecting inline requests, the body of g invokes the public
method f, so g could be inlined into t. However, once f is inlined
into g, the body of g contains a try-catch block. Now we cannot inline
g into t anymore, because the call stack at the g callsite is
non-empty (the stack is cleared when entering a handler).

lrytz committed:
Simplify post inlining requests

Clean up inliner test

lrytz committed:
Allow @inline/noinline at callsites (in addition to def-site)

Allow annotating individual callsites @inline / @noinline using an
annotation ascription

    c.foo(): @inline

lrytz commented:
review by @retronym

sjrd commented:
The last commit (callsite `@inline/@noinline`) doesn't sound like a "cleanup". It's an entirely new feature.

It also has consequences that go beyond the inliner itself. First, other back-ends could reuse the attachments to also take callsite `@inline/@noinline` into account (I've had a mind to do this in Scala.js for some time).

But more importantly, how is this going to interact with sbt's incremental compilation? Currently, if you change the body of an `@inline` method, sbt considers it a binary incompatible change to the method (even if the interface is untouched) and will recompile callsites. Now, if you don't have `@inline` at definition site, but you have `@inline` at callsite, how would sbt know to recompile the client when the callee's body is updated?

lrytz commented:
Thanks @sjrd, I was sloppy in the PR title - we should indeed discuss this a bit more.

In terms of interaction with the incremental compiler: there's noting new. Incremental compilation in sbt is incorrect when the optimizer is enabled, that is the case in 2.11.7, and we're not planning to fix this. Example:

```
➜  sbtproj  ls src/main/scala/
A.scala    B.scala    Main.scala
➜  sbtproj  cat src/main/scala/*
class A {
  @inline final def f = 3
}
class B {
  @inline final def f = (new A).f
}
object Main extends App {
  println((new B).f)
}
➜  sbtproj  cat build.sbt
scalaVersion := "2.11.7"

scalacOptions += "-optimize"
➜  sbtproj  sbt clean run
[info] Compiling 3 Scala sources to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3
```

Now I edit `A.scala`

```
➜  sbtproj  cat src/main/scala/A.scala
class A {
  @inline final def f = 3333
}
➜  sbtproj  sbt run
[info] Compiling 1 Scala source to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3
➜  sbtproj  sbt clean run
[info] Compiling 3 Scala sources to /Users/luc/tmp/sbtproj/target/scala-2.11/classes...
[info] Running Main
3333
```

sjrd commented:
> Incremental compilation in sbt is incorrect when the optimizer is enabled, that is the case in 2.11.7, and we're not planning to fix this.

Oh, OK then.

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala:
```diff
@@ -48,19 +47,7 @@ class Inliner[BT <: BTypes](val btypes: BT) {
       // need to run DCE before inlining g.
       eliminateUnreachableCodeAndUpdateCallGraph(callee.callee, callee.calleeDeclarationClass.internalName)
```
todo: move this down to `def inline`?

lrytz committed:
Correctly handle post-inline requests of non-existing callsites

A post-inline request is allowed to refer to a callsite that does not
exist anymore because it was alredy inlined while handling some other
inline request, or because it was DCE'd.

lrytz commented:
@sjrd i think the `@inline`-ascription should work for scala.js - anything you'd like to have differently from your side? Here's how I check the AST if an Apply is ascribed (the tree attachment is added during Typers): https://github.com/lrytz/scala/commit/6f72cb19b266c26e154df75af444983ca41865ec#diff-2e0fd5042aa83407a80a7557cc2a666bR710

Test cases here: https://github.com/lrytz/scala/commit/6f72cb19b266c26e154df75af444983ca41865ec#diff-c108d2950579fe9645aa47c8a1982256R1059

sjrd commented:
It looks like I could do a similar test in Scala.js, yes. I don't think it should be any different. :-)

lrytz commented:
/rebuild

lrytz commented:
green now - review by @retronym

retronym commented:
LGTM!
